src/examples/millionairs_problem/millionairs_problem.cpp

```diff
@@ -0,0 +1,177 @@
+// MIT License
+//
+// Copyright (c) 2019 Oleksandr Tkachenko, Daniel Demmler, Tobias Kussel
+// Cryptography and Privacy Engineering Group (ENCRYPTO)
+// TU Darmstadt, Germany
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include <cmath>
+#include <fstream>
+#include <iostream>
+#include <random>
+#include <regex>
+
+#include <fmt/format.h>
+#include <boost/lexical_cast.hpp>
+#include <boost/program_options.hpp>
+#include <stdexcept>
+
+#include "base/party.h"
+#include "common/millionair.h"
+#include "communication/communication_layer.h"
+#include "communication/tcp_transport.h"
+
+namespace program_options = boost::program_options;
+
+bool CheckPartyArgumentSyntax(const std::string& party_argument);
+
+std::pair<program_options::variables_map, bool> ParseProgramOptions(int ac, char* av[]);
+
+encrypto::motion::PartyPointer CreateParty(const program_options::variables_map& user_options);
+
+int main(int ac, char* av[]) {
+  auto [user_options, help_flag] = ParseProgramOptions(ac, av);
+  // if help flag is set - print allowed command line arguments and exit
+  if (help_flag) return EXIT_SUCCESS;
+
+  auto party  = CreateParty(user_options);
+  size_t my_id = user_options["my-id"].as<std::size_t>();
+  //assert(my_id >= 0 && my_id <=1);
+  uint32_t input = user_options["input"].as<std::uint32_t>();
+  std::cout << "I'm Party " << my_id << " and my input is: " << input << std::endl;
+
+  EvaluateProtocol(party, input);
+  return EXIT_SUCCESS;
+}
+
+const std::regex kPartyArgumentRegex(
+    "(\\d+),(\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}),(\\d{1,5})");
+
+bool CheckPartyArgumentSyntax(const std::string& party_argument) {
+  // other party's id, IP address, and port
+  return std::regex_match(party_argument, kPartyArgumentRegex);
```

Comment:
Can't we accept any string for the host here? The TCPTransport will complain if it cannot be resolved. This way it would also support IPv6 and other hostnames.

Reply:
This part is verbatim from the example_template. In my opinion relying on the parsing of TCPTransport is a good and more flexible solution. However that should be refactored across all examples.

Reply:
It would also be a good idea to have some object or function that all examples could use to avoid repetitions and simplify refactoring.